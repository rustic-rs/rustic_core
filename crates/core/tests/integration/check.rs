use std::{fs::File, path::Path, sync::Arc};

use anyhow::Result;
use flate2::read::GzDecoder;
use insta::assert_debug_snapshot;
use rstest::rstest;
use tar::Archive;
use tempfile::tempdir;

use rustic_backend::LocalBackend;
use rustic_core::{CheckOptions, Repository, RepositoryBackends, RepositoryOptions};

#[rstest]
#[case("repo-data-missing.tar.gz")]
#[case("repo-duplicates.tar.gz")]
#[case("repo-index-missing-blob.tar.gz")]
#[case("repo-index-missing.tar.gz")]
#[case("repo-mixed.tar.gz")]
#[case("repo-obsolete-index.tar.gz")]
#[case("repo-unreferenced-data.tar.gz")]
#[case("repo-unused-data-missing.tar.gz")]
fn test_check(#[case] repo_file: String) -> Result<()> {
    // unpack repo
    let dir = tempdir()?;
    let path = Path::new("tests/fixtures/").join(&repo_file);
    let tar_gz = File::open(path)?;
    let tar = GzDecoder::new(tar_gz);
    let mut archive = Archive::new(tar);
    archive.unpack(&dir)?;

    let be = LocalBackend::new(dir.path().join("repo").to_str().unwrap(), None)?;
    let be = RepositoryBackends::new(Arc::new(be), None);
    let options = RepositoryOptions::default().password("geheim");
    let repo = Repository::new(&options, &be)?.open()?;

    let opts = CheckOptions::default().read_data(true);
    let check_results = repo.check(opts)?;
    assert_debug_snapshot!(repo_file, check_results);

    Ok(())
}
