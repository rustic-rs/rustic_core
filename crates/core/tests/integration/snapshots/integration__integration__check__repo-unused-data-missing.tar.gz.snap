---
source: crates/core/tests/integration/check.rs
expression: check_results
---
CheckResults(
    [
        (
            Error,
            PackSizeMismatchIndex {
                id: f6a0bdd2515a5fc8d4253a4ba50abe03649d710cd2f6142837c86c760b94a43a,
                index_size: 2974,
                size: 0,
            },
        ),
        (
            Error,
            NoPack {
                id: PackId(
                    b2a0e6762aeeb8205967f326929e7f07f4997eed540fc8a759604f4baba43027,
                ),
            },
        ),
        (
            Error,
            NoPack {
                id: PackId(
                    d05c81069fdf4616aaf89ddf530dd312633176e6ca8f24fd4c878e453aa57b81,
                ),
            },
        ),
    ],
)
